tests/ci/worker/init_runner.sh

```diff
@@ -125,6 +180,20 @@ while true; do
         sleep 15
     else
         echo "Runner is working with pid $runner_pid, nothing to do"
-        sleep 10
+        # The runner does not provide a way to determine, if it runs the job,
+        # neither the way to determine if it just litens. But there should be a
+        # process for Runner.Worker. So if the runner just hangs around for long,
+        # we check if it's fine to let it go
+        if ! pgrep Runner.Worker > /dev/null; then
```

Comment:
@Felixoid I saw similar issues with this approach I found out that it's better to check the log file instead, sometimes it takes a bit from printing this message and starting the worker.

```js
async function isJobScheduled() {
    return await $`cat actions-runner/_diag/Runner*.log | grep "Running job: "`.nothrow().timeout('5s').exitCode === 0
}
```

Reply:
I see an issue here, that there many logs in the directory.

Note that we don't launch an instance per job, but reutilize the same instance for runners many times. It's too expensive for us to spawn an instance per job and generally a no-way solution.

Reply:
It makes sense for sure. Actually, between this check and the following teardown pretty long time.

Of chaos.. This part definitely requires some rework. It should check if the instance should go down first, and only then check if there's now Worker. And then quickly tear down.

Reply:
Thank you, the second pair of eyes helps a lot!

Reply:
So, this part is reworked to have less gap between the check and killing the instance.

First, the `Runner.Worker` process is checked, then I check for the metadata. And only when the metadata says, that it's time to go, I check again for the worker process, and then kill the runner with fire.

`-9` is used to kill the instance immediately without letting it mark the possibly launched jobs as `cancelled`. This status is a plague to us, and we avoid it to restart all the workflows automatically.

The issue here, `./config.sh remove` doesn't work in this case, because the runner is still "running". So maybe, in this place graceful killing would make sense.

Reply:
I ended up with the same solution, checking the `lsof $runner_pid` and getting the log file from there. I hope it will allow us to reduce the killed rate even more

Reply:
It seems to me like that too. I was thinking about additional solution to this via the runner tag removal, before shutting down the runner that is waiting for the job, first remove the tag, wait a minute and tear down the runner. It however adds another complexity to the solution, because of the github api access and might hit api rate limits at bigger scale. The other was the other way around and attach unique label to the runner, then have a job to trigger that will target the specific runner, here I am not sure if you can have unique tags per runner.

For now I try not to remove instances when jobs are running.

Reply:
> I was thinking about additional solution to this via the runner tag removal, before shutting down the runner that is waiting for the job, first remove the tag, wait a minute and tear down the runner.

Is it possible on the fly? Interesting, looks like [it is](https://docs.github.com/en/rest/actions/self-hosted-runners?apiVersion=2022-11-28#remove-all-custom-labels-from-a-self-hosted-runner-for-an-organization). As far as it is quite wild, it actually could work. The rate shouldn't be big. Wait for 1 minute first, then delete labels, and after another minute - delete the runner. If the Job would launch, well, we are good to go

> The other way around would be to attach the unique label to the runner, then have a job that will target the specific runner that would tear down after completion, here I am not sure if you can have unique tags per runner.

It's a dead end. Metaprogramming to work around fundamental issues. At this point, it's better to look around for another CI system, IMHO.

> For now I try not to remove instances when jobs are running.

Spice must flow, war never changes, CI never stops.